src/scripts/Joyride.jsx

```diff
@@ -144,29 +152,59 @@ class Joyride extends React.Component {
   }
 
   componentWillReceiveProps(nextProps) {
-    const { play, shouldPlay, standaloneTooltip } = this.state;
-    const { keyboardNavigation, run } = this.props;
     this.logger('joyride:willReceiveProps', [nextProps]);
-
-    if (this.props.steps.length && !nextProps.steps.length) {
-      this.reset();
+    const { play, shouldPlay, standaloneTooltip } = this.state;
+    const { keyboardNavigation, run, steps, stepIndex } = this.props;
+    const stepsChanged = (nextProps.steps !== steps);
+    const stepIndexChanged = (nextProps.stepIndex !== stepIndex && nextProps.stepIndex !== this.state.index);
+    const runChanged = (nextProps.run !== run);
+    let shouldStart = false;
+
+    if (stepsChanged && this.checkStepsValidity(nextProps.steps)) {
+      // Removed all steps, so reset
+      if (!nextProps.steps || !nextProps.steps.length) {
+        this.reset();
+      }
+      // Start the joyride if steps were added for the first time, and run prop is true
+      else if (!steps.length && nextProps.run) {
+        shouldStart = true;
+      }
     }
 
-    if (nextProps.steps !== this.props.steps) {
-      this.checkStepsValidity(nextProps.steps);
+    if (runChanged) {
+      // run prop was changed to off, so stop the joyride
+      if (run && nextProps.run === false) {
+        this.stop();
+      }
+      // run prop was changed to on, so start the joyride
+      else if (!run && nextProps.run) {
+        shouldStart = true;
+      }
+      // Was not playing, but should, and isn't a standaloneTooltip
+      // TODO (@ianvs): Try moving this elsewhere, as it only relies on state, not nextProps
+      else if (!play && (shouldPlay && !standaloneTooltip)) {
```

Comment:
Yeah, those are my thoughts as well.  I did not write this code, it was already here.  I _think_ it can be removed, but this PR is already large enough that it made sense to me to tackle that later in an isolated PR.